File: cmd/sample/api/metadata.py
"""
# cmd: python -m samples.metadata
# title: Metadata& Token Counting
# description: 中繼資料與 Token 計數範例
"""

import click
from .client import get_client


@click.group()
def metadata():
    """Metadata and token counting examples"""
    pass


@metadata.command()
# cmd: python -m samples.metadata with_metadata
@click.command()
def with_metadata():
    """中繼資料"""
    client = get_client()
    response = client.messages.create(
        model="MiniMax-M3",
        max_tokens=1024,
        messages=[{"role": "user", "content": "Hello!"}],
        metadata={
            "user_id": "user_123",
            "session_id": "sess_abc456",
            "agent_id": "agent_001",
            "tags": ["qa", "test"],
        },
    )
    click.echo(f"Response: {response.content[0].text}")


@metadata.command()
# cmd: python -m samples.metadata count_tokens
@click.command()
def count_tokens():
    """估算輸入 token 數量"""
    client = get_client()
    result = client.messages.count_tokens(
        model="MiniMax-M3",
        messages=[{"role": "user", "content": "Hello, how are you today?"}],
    )
    click.echo(f"Input tokens: {result.input_tokens}")


if __name__ == "__main__":
    metadata()

File: cmd/sample/api/metadata.go
package main

import (
	"context"
	"fmt"

	"github.com/anthropics/anthropic-sdk-go"
)

// Metadata 示範如何在請求中夾帶 metadata（user_id）以利後續追蹤與濫用偵測，
// 以及如何透過 CountTokens 在不發送請求的情況下預估 token 用量。
// 對應 SDK 文件: Metadata、Count tokens 章節。
func Metadata() error {
	if err := withUserMetadata(); err != nil {
		return fmt.Errorf("withUserMetadata: %w", err)
	}
	if err := countTokens(); err != nil {
		return fmt.Errorf("countTokens: %w", err)
	}
	return nil
}

// withUserMetadata — 傳遞不透明的使用者 ID（uuid / hash），不可含個資
func withUserMetadata() error {
	fmt.Println("[metadata.withUserMetadata]")
	client := getClient()
	message, err := client.Messages.New(context.TODO(), anthropic.MessageNewParams{
		Model:     defaultModel,
		MaxTokens: 256,
		Metadata: anthropic.MetadataParam{
			UserID: anthropic.String("user_abc123"),
		},
		Messages: []anthropic.MessageParam{
			anthropic.NewUserMessage(anthropic.NewTextBlock("Hello!")),
		},
	})
	if err != nil {
		return err
	}
	fmt.Printf("  usage: input=%d output=%d\n", message.Usage.InputTokens, message.Usage.OutputTokens)
	return nil
}

// countTokens — 呼叫 CountTokens 預估 input tokens，不會實際生成回應
func countTokens() error {
	fmt.Println("\n[metadata.countTokens]")
	client := getClient()
	res, err := client.Messages.CountTokens(context.TODO(), anthropic.MessageCountTokensParams{
		Model: defaultModel,
		Messages: []anthropic.MessageParam{
			anthropic.NewUserMessage(anthropic.NewTextBlock("What is the meaning of life?")),
		},
	})
	if err != nil {
		return err
	}
	fmt.Printf("  input tokens: %d\n", res.InputTokens)
	return nil
}
